File: Android/BlueChartView/src/com/blue/charts/chart/util/ChartCalculator.java
package com.blue.charts.chart.util;

import java.util.ArrayList;
import java.util.List;

import android.graphics.Paint;
import android.graphics.Rect;
import android.support.v4.view.ViewCompat;

import com.blue.charts.chart.util.ChartData.Label;
import com.blue.charts.chart.util.ChartData.Title;

/**
 * 数据处理
 * 
 * Author:chengli3209@gmail.com 2016/08/25 10:20
 */
public class ChartCalculator {
	/** 纵坐标文本矩形 */
	public Rect verticalTextRect;
	/** 横坐标文本矩形 */
	public Rect horizontalTextRect;
	/** 横坐标标题文本矩形 */
	public Rect horizontalTitleRect;

	/** 图形的高度 */
	public int height;
	/** 图形的宽度 */
	public int width;

	/** 绘制曲线图的高度 */
	public int chartHeight;

	/** 纵轴的宽度 */
	public int yAxisWidth;
	/** 纵轴的高度 */
	public int yAxisHeight;

	/** 横轴的高度 */
	public int xAxisHeight;
	/** 横轴的长度 */
	public int xAxisWidth;

	/** 横轴的标题的高度 */
	public int xTitleHeight;

	/** 灰色竖线顶点 */
	public ChartLinePoint[] gridChartLinePoints;

	/** 画布X轴的平移，用于实现曲线图的滚动效果 */
	private float translateX;

	/** 用于测量文本区域长宽的画笔 */
	private Paint paint;

	private ChartStyle style;
	private ChartData data;
	/** 光滑因子 */
	private float smoothness;

	/** 最左侧和最右侧坐标 */
	public float minX, maxX;
	public int maxRightX;
	/** 点与点之间的距离 */
	public float labelWidth;

	public ChartCalculator(ChartData data, ChartStyle style) {
		this.data = data;
		this.style = style;
		this.translateX = 0f;
		this.smoothness = 0.5f;
		this.paint = new Paint();
		this.verticalTextRect = new Rect();
		this.horizontalTextRect = new Rect();
		this.horizontalTitleRect = new Rect();
	}

	/**
	 * 计算图形绘制的参数信息
	 * 
	 * @param width
	 *            曲线图区域的宽度
	 * @param height
	 *            曲线图区域的高度
	 */
	public void compute(int width, int height) {
		this.width = width;
		this.height = height;
		this.translateX = 0;
		data.resetYLabels();
		computeAllWidthAndHeight(width, height);

		if (ChartUtils.isEmpty(data.getLineList())) {
			return;
		}
		computeVertcalAxisInfo();// 计算纵轴参数
		computeHorizontalAxisInfo();// 计算横轴参数
		computeTitlesInfo();// 计算标题参数
		computeChartLineCoordinate();// 计算纵轴参数
		computeBesselChartLinePoints();// 计算贝塞尔结点
		computeGridChartLinePoints();// 计算网格顶点
	}

	public void computeIndex() {
		if (!ChartUtils.isEmpty(data.getLineList())) {
			if (this.data.getLineList().size() <= data.getxLabelUsageSeries())
				throw new IllegalArgumentException(
						"xLabelUsageSeries should greater than seriesList.size()");

			int size = data.getLineList().get(data.getxLabelUsageSeries())
					.getPoints().size();
			if (size <= data.getxMaxLabelCount()) {
				data.setmStartIndex(0);
				data.setmEndIndex(size);
			} else {
				data.setmStartIndex(size - data.getxMaxLabelCount());
				data.setmEndIndex(data.getmStartIndex()
						+ data.getxMaxLabelCount());
			}
		}
	}

	/**
	 * 计算所有视图的宽高
	 */
	public void computeAllWidthAndHeight(int width, int height) {
		// y轴的宽度
		paint.setTextSize(style.getVerticalLabelTextSize());
		List<Label> yLabels = data.getyLabels();
		String maxText = getMaxText(yLabels);
		paint.getTextBounds(maxText, 0, maxText.length(), verticalTextRect);
		// 计算y坐标文本的宽度
		yAxisWidth = (int) (verticalTextRect.width() * 1.5f);

		// 横轴的高度
		paint.setTextSize(style.getHorizontalLabelTextSize());
		String rectText = "测";
		paint.getTextBounds(rectText, 0, rectText.length(), horizontalTextRect);
		xAxisHeight = horizontalTextRect.height() * 2;

		// 计算示例图的高度
		xTitleHeight = 0;
		if (style.isShowLegendView()) {
			paint.setTextSize(style.getHorizontalTitleTextSize());
			String titleText = data.getLineList().get(0).getLineTitle();
			paint.getTextBounds(titleText, 0, titleText.length(),
					horizontalTitleRect);
			xTitleHeight = horizontalTitleRect.height() * 2;
		}

		labelWidth = ((width - yAxisWidth) * 1.0f) / data.getxMaxLabelCount();
		xAxisWidth = width - yAxisWidth;
		maxRightX = (width - yAxisWidth) - xAxisWidth;

		// 绘制区域的高度
		chartHeight = height - xTitleHeight;
		// 计算纵轴的高度
		yAxisHeight = chartHeight - xAxisHeight;
	}

	/** 计算纵轴参数 */
	private void computeVertcalAxisInfo() {
		paint.setTextSize(style.getVerticalLabelTextSize());
		// 计算y坐标文本的宽度
		float x = yAxisWidth * 0.5f;
		float vh = verticalTextRect.height();

		List<Label> yLabels = data.getyLabels();
		int yLabelCount = yLabels.size();
		float padding = yAxisHeight / yLabelCount;

		for (int i = 0; i < yLabelCount; i++) {
			Label label = yLabels.get(i);
			label.index = i;
			label.x = x;

			if (i == 0) {
				label.y = vh;
			} else {
				label.y = padding * i;
			}
			label.drawingY = label.y + vh;
		}
	}

	/** 计算横轴参数 */
	private void computeHorizontalAxisInfo() {
		List<Label> xLabels = data.getxLabels();
		paint.setTextSize(style.getHorizontalLabelTextSize());
		// int size = xLabels.size();
		for (int i = data.getmStartIndex(); i < data.getmEndIndex(); i++) {
			Label label = xLabels.get(i);
			label.index = i;
			// label.width = paint.measureText(label.text);
			// label.height = horizontalTextRect.height();

			label.x = labelWidth * ((i - data.getmStartIndex()) + 0.3f);
			label.y = chartHeight - horizontalTextRect.height() * 0.5f;

			/*
			 * label.drawingY = label.y; if (i == 0 || (i == data.getmEndIndex()
			 * - 1 && i == size - 1)) { label.drawingX = label.x; } else {
			 * label.drawingY = label.x - label.width / 2.0f; }
			 */
		}

		minX = xLabels.get(data.getmStartIndex()).x;
		maxX = xLabels.get(data.getmEndIndex() - 1).x;
	}

	/** 计算标题的坐标信息 */
	private void computeTitlesInfo() {
		if (style.isShowLegendView()) {
			paint.setTextSize(style.getHorizontalTitleTextSize());

			List<Title> titles = data.getTitles();
			int count = titles.size();
			float stepX = (width - style.getHorizontalTitlePaddingLeft() - style
					.getHorizontalTitlePaddingRight()) / count;
			for (Title title : titles) {
				title.radius = 10;
				title.circleTextPadding = 10;
				title.updateTextRect(paint, stepX);
				title.textX = style.getHorizontalTitlePaddingLeft()
						+ (titles.indexOf(title) + 0.5f) * stepX;
				title.textY = xTitleHeight * 0.75f;
				title.circleX = title.textX - title.textRect.width() / 2
						- title.circleTextPadding - title.radius;
				title.circleY = title.textY - horizontalTitleRect.height()
						* 0.5f + 5;
			}
		}
	}

	/** 计算序列的坐标信息 */
	private void computeChartLineCoordinate() {
		List<Label> yLabels = data.getyLabels();
		float minCoordinateY = yLabels.get(0).y;
		float maxCoordinateY = yLabels.get(yLabels.size() - 1).y;

		for (ChartLine chartLine : data.getLineList()) {
			List<ChartLinePoint> chartLinePoints = chartLine.getPoints();
			// float chartLinePointWidth = xAxisWidth / chartLinePoints.size();
			float chartLinePointWidth = labelWidth;
			for (int i = data.getmStartIndex(); i < data.getmEndIndex(); i++) {
				ChartLinePoint chartLinePoint = chartLinePoints.get(i);
				// 计算数据点的坐标
				chartLinePoint.x = chartLinePointWidth
						* ((i - data.getmStartIndex()) + 0.3f);
				float ratio = (chartLinePoint.getyValue() - data.getMinValueY())
						/ (float) (data.getMaxValueY() - data.getMinValueY());
				chartLinePoint.y = maxCoordinateY
						- (maxCoordinateY - minCoordinateY) * ratio;
			}
		}
	}

	/**
	 * 获取label中最长的文本
	 * 
	 * @param labels
	 * @return
	 */
	private String getMaxText(List<Label> labels) {
		String maxText = "";
		for (Label label : labels) {
			if (label.text.length() > maxText.length())
				maxText = label.text;
		}
		return maxText;
	}

	/** 计算网格顶点 */
	private void computeGridChartLinePoints() {
		gridChartLinePoints = new ChartLinePoint[data.getMaxPointsCount()];
		List<ChartLine> chartLineList = data.getLineList();
		for (ChartLine chartLine : chartLineList) {
			for (int i = data.getmStartIndex(); i < data.getmEndIndex(); i++) {
				ChartLinePoint chartLinePoint = chartLine.getPoints().get(i);
				int index = chartLine.getPoints().indexOf(chartLinePoint);
				if (gridChartLinePoints[index] == null
						|| gridChartLinePoints[index].getyValue() < chartLinePoint
								.getyValue()) {
					gridChartLinePoints[index] = chartLinePoint;
				}
			}
		}
	}

	/** 计算贝塞尔结点 */
	private void computeBesselChartLinePoints() {
		for (ChartLine chartLine : data.getLineList()) {
			List<ChartLinePoint> besselChartLinePoints = chartLine
					.getBesselPoints();
			List<ChartLinePoint> chartLinePoints = new ArrayList<ChartLinePoint>();
			for (int i = data.getmStartIndex(); i < data.getmEndIndex(); i++) {
				ChartLinePoint chartLinePoint = chartLine.getPoints().get(i);
				if (chartLinePoint.getyValue() > 0)
					chartLinePoints.add(chartLinePoint);
			}

			int count = chartLinePoints.size();
			if (count < 2)
				continue;

			besselChartLinePoints.clear();
			for (int i = 0; i < count; i++) {
				if (i == 0 || i == count - 1) {
					computeUnMonotoneChartLinePoints(i, chartLinePoints,
							besselChartLinePoints);
				} else {
					ChartLinePoint p0 = chartLinePoints.get(i - 1);
					ChartLinePoint p1 = chartLinePoints.get(i);
					ChartLinePoint p2 = chartLinePoints.get(i + 1);
					if ((p1.y - p0.y) * (p1.y - p2.y) >= 0) {// 极值点
						computeUnMonotoneChartLinePoints(i, chartLinePoints,
								besselChartLinePoints);
					} else {
						computeMonotoneChartLinePoints(i, chartLinePoints,
								besselChartLinePoints);
					}
				}
			}
		}
	}

	/** 计算非单调情况的贝塞尔结点 */
	private void computeUnMonotoneChartLinePoints(int i,
			List<ChartLinePoint> chartLinePoints,
			List<ChartLinePoint> besselChartLinePoints) {
		if (i == 0) {
			ChartLinePoint p1 = chartLinePoints.get(0);
			ChartLinePoint p2 = chartLinePoints.get(1);
			besselChartLinePoints.add(p1);
			besselChartLinePoints.add(new ChartLinePoint(p1.x + (p2.x - p1.x)
					* smoothness, p1.y));
		} else if (i == chartLinePoints.size() - 1) {
			ChartLinePoint p0 = chartLinePoints.get(i - 1);
			ChartLinePoint p1 = chartLinePoints.get(i);
			besselChartLinePoints.add(new ChartLinePoint(p1.x - (p1.x - p0.x)
					* smoothness, p1.y));
			besselChartLinePoints.add(p1);
		} else {
			ChartLinePoint p0 = chartLinePoints.get(i - 1);
			ChartLinePoint p1 = chartLinePoints.get(i);
			ChartLinePoint p2 = chartLinePoints.get(i + 1);
			besselChartLinePoints.add(new ChartLinePoint(p1.x - (p1.x - p0.x)
					* smoothness, p1.y));
			besselChartLinePoints.add(p1);
			besselChartLinePoints.add(new ChartLinePoint(p1.x + (p2.x - p1.x)
					* smoothness, p1.y));
		}
	}

	/**
	 * 计算单调情况的贝塞尔结点
	 * 
	 * @param i
	 * @param ChartLinePoints
	 * @param besselChartLinePoints
	 */
	private void computeMonotoneChartLinePoints(int i,
			List<ChartLinePoint> ChartLinePoints,
			List<ChartLinePoint> besselChartLinePoints) {
		ChartLinePoint p0 = ChartLinePoints.get(i - 1);
		ChartLinePoint p1 = ChartLinePoints.get(i);
		ChartLinePoint p2 = ChartLinePoints.get(i + 1);
		float k = (p2.y - p0.y) / (p2.x - p0.x);
		float b = p1.y - k * p1.x;
		ChartLinePoint p01 = new ChartLinePoint();
		p01.x = p1.x - (p1.x - (p0.y - b) / k) * smoothness;
		p01.y = k * p01.x + b;
		besselChartLinePoints.add(p01);
		besselChartLinePoints.add(p1);
		ChartLinePoint p11 = new ChartLinePoint();
		p11.x = p1.x + (p2.x - p1.x) * smoothness;
		p11.y = k * p11.x + b;
		besselChartLinePoints.add(p11);
	}

	public void setSmoothness(float smoothness) {
		this.smoothness = smoothness;
	}

	/**
	 * 计算移动在当前哪个point
	 * 
	 * @param moveX
	 */
	public Label computeMoveIndex(float moveX) {
		if (labelWidth <= 0) {
			return null;
		}
		int index = (int) (moveX / labelWidth) + data.getmStartIndex();
		int size = data.getxLabels().size();
		if (size > 0 && index >= 0 && index < size) {
			return data.getxLabels().get(index);
		}
		return null;
	}

	/**
	 * 平移画布
	 * 
	 * @param distanceX
	 * @param velocityX
	 */
	public void move(float distanceX) {
		int moveIndex = (int) (Math.abs(distanceX) / labelWidth);
		if (moveIndex <= 0) {
			moveIndex = 1;
		} else if (moveIndex > 1) {
			moveIndex = 1;
		}
		if (ChartUtils.isEmpty(data.getLineList())) {
			return;
		}
		int size = data.getLineList().get(0).getPoints().size();
		if (size <= data.getxMaxLabelCount()) {
			return;
		}

		int start = data.getmStartIndex();
		int end = data.getmEndIndex();
		if (distanceX > 0) {
			start = start + moveIndex;
		} else {
			start = start - moveIndex;
			if (start <= 0) {
				start = 0;
			}
		}

		if (start >= size - data.getxMaxLabelCount()) {
			start = size - data.getxMaxLabelCount();
		}
		end = start + data.getxMaxLabelCount();

		data.setmStartIndex(start);
		data.setmEndIndex(end);
		compute(width, height);
		ViewCompat
				.postInvalidateOnAnimation(data.chartViewLayout.chartVerticalAxes);
		translateX = translateX - distanceX;
	}

	/**
	 * 平移画布
	 * 
	 * @param distanceX
	 * @param velocityX
	 */
	public void moveTo(int x) {
		translateX = x;
	}

	public float getTranslateX() {
		return translateX;
	}

	/***
	 * 确保画布的移动不会超出范围
	 * 
	 * @return true,超出范围；false，未超出范围
	 */
	public boolean ensureTranslation() {
		if (translateX >= 0) {
			translateX = 0;
			return true;
		} else if (translateX < 0) {
			if (yAxisWidth != 0 && translateX < maxRightX) {
				translateX = maxRightX;
				return true;
			}
		}

		ChartUtils.log("ensureTranslation", "translateX = " + translateX
				+ ",maxRightX = " + maxRightX);
		return false;
	}
}

File: IOS/BlueChartView/BlueChartView/ChartCalculator.m
//
//  ChartCalculator.m
//  BlueChartView
//
//  Created by blue on 16/8/30.
//  Copyright © 2016年 chengli. All rights reserved.
//

#import "ChartCalculator.h"

@implementation ChartCalculator

-(instancetype)initWith:(ChartData *)data style:(ChartStyle *)style {
    self = [super init];
    
    if(self){
        self.data = data;
        self.style = style;
        
        self.translateX = 0.0f;
        self.smoothness = 0.5f;
        
        self.gridChartLinePoints = [NSMutableArray new];
    }
    
    return self;
}

#pragma 计算图形绘制的参数信息
-(void)compute:(CGFloat)width height:(CGFloat)height{
    self.width = width;
    self.height = height;
    self.translateX = 0;
    
    [self.data resetYLabels];
    [self computeAllWidthAndHeight:width height:height];
    
    if([ChartUtils isEmpty:self.data.lineList]){
        return;
    }

    [self computeVerticalAxisInfo];     //计算纵轴参数
    [self computeHorizontalAxisInfo];   //计算横轴参数
    [self computeTitlesInfo];
    [self computeChartLineCoordinate];
    [self computeBesselChartLinePoints];
    [self computeGridChartLinePoints];
}

//初始化的时候计算其索引值
-(void)computeIndex{
    if(![ChartUtils isEmpty:self.data.lineList]){
        if(self.data.lineList.count <= self.data.xLabelUsageSeries){
            return;
        }
        
        ChartLine *chartLine = [self.data.lineList objectAtIndex:self.data.xLabelUsageSeries];
        int count = (short)chartLine.points.count;
        
        if(count <= self.data.xMaxLabelCount){
            self.data.mStartIndex = 0;
            self.data.mEndIndex = count;
        }else{
            self.data.mStartIndex = (short)(count - self.data.xMaxLabelCount);
            self.data.mEndIndex = (short)(self.data.mStartIndex + self.data.xMaxLabelCount);
        }
        
        NSLog(@"start = %d,end = %d",self.data.mStartIndex,self.data.mEndIndex);
    }
}

//计算绘制试图的宽高
-(void)computeAllWidthAndHeight:(CGFloat)width height:(CGFloat)height{
    // y轴的宽度
    NSString *maxText = [self getMaxLengthText];
    self.yAxisWidth = [ChartUtils widthForString:maxText fontSize:self.style.verticalLabelTextSize] * 1.5f;
    
    // x轴的高度
    NSString *text = @"测";
    self.xAxisHeight = [ChartUtils heightForString:text fontSize:self.style.horizontalLabelTextSize] * 1.5f;
    
    // 计算示例图的高度
    self.xTitleHeight = 0;
    if(self.style.isShowLegendView){
        ChartLine *line = [self.data.lineList objectAtIndex:0];
        text = line.lineTitle;
        self.xTitleHeight = [ChartUtils heightForString:text fontSize:self.style.horizontalTitleTextSize] * 2.0f;
    }
    
    //点之间间距
    self.labelWidth = (width - self.yAxisWidth)/(self.data.xMaxLabelCount - 1);
    //最右侧的x坐标
    self.maxRightX = (width - self.yAxisWidth) - self.xAxisWidth;
    
    //绘制曲线区域的宽度
    self.xAxisWidth = (width - self.yAxisWidth);
    //绘制曲线区域的高度
    self.chartHeight = height - self.xTitleHeight;
    //绘制纵轴的高度
    self.yAxisHeight = self.chartHeight - self.xAxisHeight;
}

//获取长度最长的字符串
-(NSString *)getMaxLengthText{
    NSString *max = @"";
    
    for(Label *label in self.data.yLabels) {
        if([label.text length] > [max length]){
            max = label.text;
        }
    }
    
    return max;
}

//计算纵轴参数
-(void)computeVerticalAxisInfo{
    CGFloat vh = [ChartUtils heightForString:@"测" fontSize:self.style.verticalLabelTextSize];
    
    NSMutableArray *yLabels = self.data.yLabels;
    NSInteger yLableCount = [yLabels count];
    CGFloat padding = self.yAxisHeight / yLableCount;
    
    for(int i = 0; i < yLableCount; i++){
        Label *label = [yLabels objectAtIndex:i];
        label.index = i;
        CGFloat w = [ChartUtils widthForString:label.text fontSize:self.style.verticalLabelTextSize];
        label.width = w;
        label.height = vh;
        
        label.x = self.yAxisWidth - w;
        
        if(i == 0){
            label.y = 0;
        }else{
            label.y = padding * i;
        }
        label.drawingY = label.y + vh;
    }
}

//计算横轴参数
-(void)computeHorizontalAxisInfo{
    NSMutableArray *xLabels = self.data.xLabels;
    CGFloat vh = [ChartUtils heightForString:@"测" fontSize:self.style.horizontalLabelTextSize];
    NSInteger count = xLabels.count;
    
    for(int i = self.data.mStartIndex; i < self.data.mEndIndex;i++){
        Label *label = [xLabels objectAtIndex:i];
        label.index = i;
        label.width = [ChartUtils widthForString:label.text fontSize:self.style.verticalLabelTextSize];;
        label.height = vh;
        
        label.x = self.labelWidth * ((i - self.data.mStartIndex));
        if(i == self.data.mStartIndex){
            label.x += self.style.radius;
        }else if(i == self.data.mEndIndex - 1){
            label.x -= self.style.radius;
        }
        
        label.y = self.yAxisHeight;
        
        if(i == 0 || (i == self.data.mEndIndex - 1 && self.data.mEndIndex == count - 1)){
            label.drawingX = label.x;
        }else{
            label.drawingX = label.x - (label.width) / 2.0f;
        }
        label.drawingY = self.yAxisHeight;
        
        if(i == self.data.mStartIndex){
            self.minX = label.x;//最小值
        }else if(i == self.data.mEndIndex - 1){
            self.maxX = label.x;//最大值
        }
    }
}

//计算标题的坐标
-(void)computeTitlesInfo{
    if(self.style.isShowLegendView){
        NSMutableArray *titles = self.data.titles;
        NSInteger count = [titles count];
        CGFloat stepX = (self.width - self.style.horizontalTitlePaddingLeft - self.style.horizontalTitlePaddingRight)/count;
    
        for(Title *title in titles){
            title.radius = 10.0f;
            title.circleTextPadding = 10.0f;
            [title updateTextRect:self.style.horizontalTitleTextSize maxWidth:stepX];
            title.textX = self.style.horizontalTitlePaddingLeft
            + ([titles indexOfObject:title] + 0.5f) * stepX;
            title.textY = self.xTitleHeight * 0.75f;
            
            title.circleX = title.textX - title.width / 2 - title.circleTextPadding - title.radius;
            title.circleY = title.textY - title.height * 0.5f + 5;
        }
    }
}

//计算序列坐标信息
-(void)computeChartLineCoordinate{
    NSMutableArray *yLabels = self.data.yLabels;
    CGFloat minCoordinateY = ((Label *)[yLabels objectAtIndex:0]).y;
    CGFloat maxCoordinateY = ((Label *)[yLabels objectAtIndex:yLabels.count - 1]).y;
    
    for(ChartLine *chartLine in self.data.lineList){
        float pointWidth = self.labelWidth;
        for(int i = self.data.mStartIndex;i < self.data.mEndIndex;i++){
            ChartLinePoint *point = [chartLine.points objectAtIndex:i];
            point.x = pointWidth * ((i - self.data.mStartIndex));
            if(i == self.data.mStartIndex){
                point.x += self.style.radius;
            }else if(i == self.data.mEndIndex - 1){
                point.x -= self.style.radius;
            }
            float ratio = (point.yValue - self.data.minValueY) / (self.data.maxValueY - self.data.minValueY);
            point.y = maxCoordinateY - (maxCoordinateY - minCoordinateY) * ratio;
        }
    }
}

//计算贝塞尔节点
-(void)computeBesselChartLinePoints{
    for (ChartLine *chartLine in self.data.lineList) {
        NSMutableArray *besselChartLinePoints = chartLine.besselPoints;
        NSMutableArray *chartLinePoints = [NSMutableArray new];
        
        for (int i = self.data.mStartIndex; i < self.data.mEndIndex; i++) {
            ChartLinePoint *chartLinePoint = [chartLine.points objectAtIndex:i];
            if (chartLinePoint.yValue > 0){
                [chartLinePoints addObject:chartLinePoint];
            }
        }
        
        NSInteger count = chartLinePoints.count;
        if (count < 2)
            continue;
        
        [besselChartLinePoints removeAllObjects];
        for (int i = 0; i < count; i++) {
            if (i == 0 || i == count - 1) {
                [self computeUnMonotoneChartLinePoints:i points:chartLinePoints besselPoints:besselChartLinePoints];
            } else {
                ChartLinePoint *p0 = [chartLinePoints objectAtIndex:i - 1];
                ChartLinePoint *p1 = [chartLinePoints objectAtIndex:i];
                ChartLinePoint *p2 = [chartLinePoints objectAtIndex:i + 1];
                if ((p1.y - p0.y) * (p1.y - p2.y) >= 0) {// 极值点
                    [self computeUnMonotoneChartLinePoints:i points:chartLinePoints besselPoints:besselChartLinePoints];
                } else {
                    [self computeMonotoneChartLinePoints:i points:chartLinePoints besselPoints:besselChartLinePoints];
                }
            }
        }
    }
}

/** 计算非单调情况的贝塞尔结点 */
-(void)computeUnMonotoneChartLinePoints:(NSInteger)i points:(NSMutableArray *)chartLinePoints besselPoints:(NSMutableArray *)besselChartLinePoints {
    if (i == 0) {
        ChartLinePoint *p1 = [chartLinePoints objectAtIndex:0];
        ChartLinePoint *p2 = [chartLinePoints objectAtIndex:1];
        
        [besselChartLinePoints addObject:p1];
        [besselChartLinePoints addObject:[[ChartLinePoint alloc]initXY:p1.x + (p2.x - p1.x)
                                          * self.smoothness y:p1.y drawPoint:NO]];
    } else if (i == chartLinePoints.count - 1) {
        ChartLinePoint *p0 = [chartLinePoints objectAtIndex:i - 1];
        ChartLinePoint *p1 = [chartLinePoints objectAtIndex:i];
        
        [besselChartLinePoints addObject:[[ChartLinePoint alloc]initXY:p1.x - (p1.x - p0.x)
                                           * self.smoothness y:p1.y drawPoint:NO]];
        [besselChartLinePoints addObject:p1];
    } else {
        ChartLinePoint *p0 = [chartLinePoints objectAtIndex:i - 1];
        ChartLinePoint *p1 = [chartLinePoints objectAtIndex:i];
        ChartLinePoint *p2 = [chartLinePoints objectAtIndex:i + 1];
        
        [besselChartLinePoints addObject:[[ChartLinePoint alloc]initXY:p1.x - (p1.x - p0.x)
                                           * self.smoothness y:p1.y drawPoint:NO]];
        [besselChartLinePoints addObject:p1];
        [besselChartLinePoints addObject:[[ChartLinePoint alloc]initXY:p1.x + (p2.x - p1.x)
                                           * self.smoothness y:p1.y drawPoint:NO]];
    }
}

//计算单调情况的贝塞尔结点
-(void) computeMonotoneChartLinePoints:(NSInteger)i points:(NSMutableArray *)chartLinePoints besselPoints:(NSMutableArray*)besselChartLinePoints {
    ChartLinePoint *p0 = [chartLinePoints objectAtIndex:i - 1];
    ChartLinePoint *p1 = [chartLinePoints objectAtIndex:i];
    ChartLinePoint *p2 = [chartLinePoints objectAtIndex:i + 1];
    float k = (p2.y - p0.y) / (p2.x - p0.x);
    float b = p1.y - k * p1.x;
    ChartLinePoint *p01 = [ChartLinePoint new];
    p01.x = p1.x - (p1.x - (p0.y - b) / k) * self.smoothness;
    p01.y = k * p01.x + b;
    [besselChartLinePoints addObject:p01];
    [besselChartLinePoints addObject:p1];
    ChartLinePoint *p11 = [ChartLinePoint new];
    p11.x = p1.x + (p2.x - p1.x) * self.smoothness;
    p11.y = k * p11.x + b;
    [besselChartLinePoints addObject:p11];
}

//计算网格顶点
-(void)computeGridChartLinePoints{
    [self.gridChartLinePoints removeAllObjects];
    
    NSMutableArray *chartLineList = self.data.lineList;
    for(ChartLine *chartLine in chartLineList){
        for(int i = self.data.mStartIndex ;i < self.data.mEndIndex;i++){
            ChartLinePoint *point = [chartLine.points objectAtIndex:i];
            //NSInteger index = [chartLine.points indexOfObject:point];
            [self.gridChartLinePoints addObject:point];
        }
    }
}

//计算当前移动的点在哪个区域索引Label
-(Label *)computeMoveIndex:(CGFloat)moveX{
    if(self.labelWidth <= 0){
        return nil;
    }
    
    NSInteger index = (int)(moveX / self.labelWidth) + self.data.mStartIndex;
    NSInteger count = self.data.xLabels.count;
    if(count > 0 && index >= 0 && index < count){
        return [self.data.xLabels objectAtIndex:index];
    }
    return nil;
}

//计算平移画布
-(void)move:(CGFloat)distanceX{
    NSInteger moveIndex = (int)(fabs(distanceX) / self.labelWidth);
    if(moveIndex <= 0){
        moveIndex = 1;
    }else if(moveIndex > 1){
        moveIndex = 1;
    }
    
    if([ChartUtils isEmpty:self.data.lineList]){
        return;
    }
    
    ChartLine *line = [self.data.lineList objectAtIndex:0];
    NSInteger count = line.points.count;
    if(count <= self.data.xMaxLabelCount){
        return;
    }
    
    NSInteger start = self.data.mStartIndex;
    NSInteger end = self.data.mEndIndex;
    
    if(distanceX < 0){
        start = start + moveIndex;
    }else {
        start = start - moveIndex;
        if(start <= 0){
            start = 0;
        }
    }
    
    if(start >= count - self.data.xMaxLabelCount){
        start = count - self.data.xMaxLabelCount;
    }
    end = start + self.data.xMaxLabelCount;
    
    self.data.mStartIndex = (short)start;
    self.data.mEndIndex = (short)end;
    [self compute:self.width height:self.height];
}

@end
